examples/gno.land/p/nt/poa/poa.gno

```diff
@@ -0,0 +1,106 @@
+package poa
+
+import (
+	"errors"
+	"std"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/sys/validators"
+)
+
+var ErrInvalidVotingPower = errors.New("invalid voting power")
+
+// PoA specifies the Proof of Authority validator set, with simple add / remove constraints.
+//
+// To add:
+// - proposed validator must not be part of the set already
+// - proposed validator voting power must be > 0
+//
+// To remove:
+// - proposed validator must be part of the set already
+type PoA struct {
+	validators *avl.Tree // std.Address -> validators.Validator
+}
+
+// NewPoA creates a new empty Proof of Authority validator set
+func NewPoA(opts ...Option) *PoA {
+	// Create the empty set
+	p := &PoA{
+		validators: avl.NewTree(),
+	}
+
+	// Apply the options
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	return p
+}
+
+func (p *PoA) AddValidator(address std.Address, pubKey string, power uint64) (*validators.Validator, error) {
+	// Validate that the operation is a valid call.
+	// Check if the validator is already in the set
+	if p.IsValidator(address) {
+		return nil, validators.ErrValidatorExists
+	}
+
+	// Make sure the voting power > 0
+	if power == 0 {
+		return nil, ErrInvalidVotingPower
+	}
+
+	v := validators.Validator{
+		Address:     address,
+		PubKey:      pubKey, // TODO: in the future, verify the public key
+		VotingPower: power,
+	}
+
+	// Add the validator to the set
+	p.validators.Set(address.String(), v)
+
+	return &v, nil
```

Comment:
If the goal is to allow the caller to update the validator object by returning a pointer, then I suggest storing a pointer as well. Alternatively, you could store the value and return a copy, but avoid a mixed approach.